src/adapter/src/coord/timeline.rs

```diff
@@ -481,65 +490,102 @@ impl Coordinator {
         global_timelines.get_mut(timeline).expect("inserted above")
     }
 
-    pub(crate) fn remove_storage_ids_from_timeline<I>(&mut self, ids: I) -> Vec<Timeline>
-    where
-        I: IntoIterator<Item = GlobalId>,
-    {
-        let mut empty_timelines = Vec::new();
-        for id in ids {
-            if let TimelineContext::TimelineDependent(timeline) = self.get_timeline_context(id) {
+    pub(crate) fn associate_with_timelines(
+        &self,
+        storage_ids: impl IntoIterator<Item = GlobalId> + Clone,
+        compute_ids: impl IntoIterator<Item = (ComputeInstanceId, GlobalId)> + Clone,
+        clusters: impl IntoIterator<Item = ComputeInstanceId> + Clone,
+    ) -> BTreeMap<Timeline, (bool, CollectionIdBundle)> {
```

Comment:
One consideration is that the bool returned doesn't really make sense in a general context. It only makes sense if this is called in the context of deleting IDs from a timeline, which isn't really conveyed anywhere and is only applicable when called from `catalog_transact`.

One thought is that this should be split up into two methods: 
1. Converts the following into a `CollectionIdBundle`:
```
storage_ids: impl IntoIterator<Item = GlobalId> + Clone,
compute_ids: impl IntoIterator<Item = (ComputeInstanceId, GlobalId)> + Clone,
clusters: impl IntoIterator<Item = ComputeInstanceId> + Clone,
```
2. Determine which timelines will be empty based on a `CollectionIdBundle` that we are going to delete. This can even probably be inlined in `catalog_transact`.